### PR TITLE
Add 'ty_error_with_guaranteed' and 'const_error_with_guaranteed'

### PR DESCRIPTION
**Pull request by chenyukang**
Part of #103874

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/expr.rs [hunk 1]
```diff
@@ -1097,12 +1096,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
             // If the assignment expression itself is ill-formed, don't
             // bother emitting another error
-            if lhs_ty.references_error() || rhs_ty.references_error() {
+            let reported = if lhs_ty.references_error() || rhs_ty.references_error() {
                 err.delay_as_bug()
             } else {
-                err.emit();
-            }
-            return self.tcx.ty_error();
+                err.emit()
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/expr.rs:1102]: `err.emit_unless` can be used here :thinking:

**Review by lcnr — COMMENTED** [review]
r=me after nit + green perf

File: compiler/rustc_hir_typeck/src/expr.rs (hunk 1 above)
Comment (chenyukang) [reply compiler/rustc_hir_typeck/src/expr.rs:1102]: change to 
```rust
 let reported = err.emit_unless(lhs_ty.references_error() || rhs_ty.references_error());
 return self.tcx.ty_error_with_guaranteed(reported);
```

**Comment by lcnr** [comment]
@rust-timer queue https://github.com/rust-lang/rust/commit/622859bba9ca3cca1ca580f819ca26e4ac9253bb

**Comment by chenyukang** [comment]
> The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3409479436/jobs/5671301865) [(plain)](https://github.com/rust-lang-ci/rust/commit/1d4e4a572b5249bf76ab7beff62d639bade9b9eb/checks/9332592042/logs)
> 
> Click to see the possible cause of the failure (guessed by this bot)
> ```
>    Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
> error[E0061]: this function takes 2 arguments but 1 argument was supplied
>     --> compiler/rustc_middle/src/ty/context.rs:1313:14
>      |
> 1313 |         self.mk_const(ty::ConstS { kind: ty::ConstKind::Error(reported), ty })
>      |                      ||
>      |                      ||
>      |                      |expected enum `kind::ConstKind`, found struct `ConstS`
>      |
> note: associated function defined here
>     --> compiler/rustc_middle/src/ty/context.rs:2589:12
>      |
>      |
> 2589 |     pub fn mk_const(self, kind: ty::ConstKind<'tcx>, ty: Ty<'tcx>) -> Const<'tcx> {
> help: provide the argument
>      |
>      |
> 1313 |         self.mk_const(/* kind::ConstKind<'_> */, /* ty::Ty<'_> */)
> 
> [RUSTC-TIMING] rustc_ast_passes test:false 4.110
> For more information about this error, try `rustc --explain E0061`.
> [RUSTC-TIMING] rustc_middle test:false 7.533
> ```

@lcnr why this error didn't happen in CI, I also can not reproduce it in local.

**Comment by lcnr** [comment]
probably you have to first pull the new changes to `master` to fix this:

```
git fetch origin
git rebase origin/master
```
(or however you're calling the rust-lang remote)

**Comment by chenyukang** [comment]
Got it, rebased master, `mk_const` is changed in upstream.

**Pull request merged**